### PR TITLE
Gcw 2189 sms refactoring bug fixes

### PR DESCRIPTION
**Pull request by steveyken**
Rewrote the code to conform to the following specifications:

Send 1 reminder if:

- recipient is donor AND
- messages have arrived since we last sent a reminder AND
- at least one message is still unread AND
- no other reminders have been sent to this user in the last 4 hours

Don't send a reminder if:

- no new message has been created since we last sent a reminder OR
- a new message has been created but it's less than 4 hours since the user last received a reminder OR
- the new message has been read already OR
- it was sent by the user

More generally

- Only ever alert for a message once
- If there is more than one new message, only send one SMS reminder.
- For users that have just signed up, don't message them for at least 4 hours

### CODE REVIEW

File: spec/services/subscriptions_reminder_spec.rb [hunk 1]
```diff
@@ -1,31 +1,109 @@
...
+        user_with_offer.update_attribute(:sms_reminder_sent_at, before_delta.ago)
+        expect(subject.send(:user_candidates_for_reminder).to_a).to eql([])
+      end
+      it "no new messages created since we last reminded them" do
+        user_with_offer.update_attribute(:sms_reminder_sent_at, before_delta.ago)
+        message.update_attribute(:created_at, (before_delta + 1).ago)
+        expect(subject.send(:user_candidates_for_reminder).to_a).to eql([])
+      end
+      it "user is not a donor (has no offers)" do
+        Offer.all.each{|offer| offer.update_attribute(:created_by_id, nil) }
```
Comment (namrataukirde) [inline spec/services/subscriptions_reminder_spec.rb]: How about Offer.update_all(created_at: nil) here ?

File: app/services/subscriptions_reminder.rb [hunk 2]
```diff
@@ -1,26 +1,40 @@
...
+  #   haven't been reminded in last X hours
+  #   have unread messages
+  #   are donors
+  #   aren't the author of the message
+  # If sms_reminder_sent_at is NULL then use created_at so we don't SMS user immediately
+  def user_candidates_for_reminder
+    user_ids = Offer.distinct.pluck(:created_by_id)
+    User.joins(subscriptions: [:message]).
+      where('users.id IN (?)', user_ids).
+      where("COALESCE(users.sms_reminder_sent_at, users.created_at) < (?)", delta.to_s(:db)).
```
[the remark this answers is not in the document]
Comment (steveyken) [reply app/services/subscriptions_reminder.rb]: hi @namrataukirde the `users.created_at` covers the scenario where the user has just signed up and sms_reminder_sent_at is `NULL`. If we didn't COALESCE here, then we'd be comparing `NULL` with a timestamp which returns 0 results.

File: lib/tasks/subscription_sms_sent_reminder_at.rake [hunk 3]
```diff
@@ -1,19 +1,6 @@
...
-
-    nil_sms_reminder_count = Subscription.where(sms_reminder_sent_at: nil).count
-    log.info("\nSubscription with nil value: #{nil_sms_reminder_count}")
-
-    Subscription.find_each(batch_size: 100) do |subscription|
-      success_count +=1 if subscription.update_column(:sms_reminder_sent_at, Time.zone.now + 24.hours)
-    end
-
-    log.info("\nTotal Records Updated: #{success_count}")
+    User.all.update(sms_reminder_sent_at: Time.zone.now)
```
[the remark this answers is not in the document]
Comment (steveyken) [reply lib/tasks/subscription_sms_sent_reminder_at.rake]: Good catch. `User.update_all` would be better

File: spec/factories/users.rb [hunk 4]
```diff
@@ -190,9 +190,9 @@
       end
     end
 
-    trait :with_organisation do
```
[the remark this answers is not in the document]
Comment (steveyken) [reply spec/factories/users.rb]: Let me check, I thought I was removing a duplicate definition. we still need it.

File: spec/services/subscriptions_reminder_spec.rb [hunk 5]
```diff
@@ -1,31 +1,109 @@
...
-  context "generate" do
-    it "sends sms of unread messages if sms_reminder_sent_at is nil" do
-      subscription_reminder.generate
-      expect(subscription.reload.sms_reminder_sent_at.to_s(:db)).to eq(time.to_s(:db))
+  context "user_candidates_for_reminder" do
+
+    context "includes user when" do
+      it "there is a new unread message and we last reminded the user over X hours ago " do
+        expect(user_with_offer.subscriptions.unread.first.message.created_at).to be > delta.ago
+        user_with_offer.update_attribute(:sms_reminder_sent_at, before_delta.ago)
```
[the remark this answers is not in the document]
Comment (steveyken) [reply spec/services/subscriptions_reminder_spec.rb]: Good to know. I'm still stuck in Rails 2/3! Always good to future-proof work. Will change this.

File: spec/services/subscriptions_reminder_spec.rb (hunk 1 above)
Comment (steveyken) [reply spec/services/subscriptions_reminder_spec.rb]: yes much better

File: app/services/subscriptions_reminder.rb (hunk 2 above)
[the remark this answers is not in the document]
Comment (namrataukirde) [reply app/services/subscriptions_reminder.rb]: Ohh Okay, Thanks 👍

[the remark this answers is not in the document]
Comment (namrataukirde) [reply app/services/subscriptions_reminder.rb]: Also I would like to know about other comment I have added regarding 4.hours.

File: spec/factories/users.rb (hunk 4 above)
[the remark this answers is not in the document]
Comment (steveyken) [reply spec/factories/users.rb]: Confirming I was just deleting a dup

File: lib/tasks/subscription_sms_sent_reminder_at.rake (hunk 3 above)
[the remark this answers is not in the document]
Comment (steveyken) [reply lib/tasks/subscription_sms_sent_reminder_at.rake]: Changed

File: spec/services/subscriptions_reminder_spec.rb (hunk 1 above)
Comment (steveyken) [reply spec/services/subscriptions_reminder_spec.rb]: fixed

Comment (namrataukirde) [reply spec/services/subscriptions_reminder_spec.rb]: :+1:

File: spec/services/subscriptions_reminder_spec.rb (hunk 5 above)
[the remark this answers is not in the document]
Comment (steveyken) [reply spec/services/subscriptions_reminder_spec.rb]: switched to update and update_column depending on callbacks and serialization required

File: app/services/subscriptions_reminder.rb (hunk 2 above)
[the remark this answers is not in the document]
Comment (steveyken) [reply app/services/subscriptions_reminder.rb]: i'll check the to_s(:db), the main thing is to ensure we are sending a TZ string, which this is not

[the remark this answers is not in the document]
Comment (steveyken) [reply app/services/subscriptions_reminder.rb]: Just to confirm, it appears when doing time comparision on `Timestamp with Timezone` fields in SQL, it is important to compare times with a TZ string. So `to_s(:db)` will not work anymore.

I'm switching to `iso8601` as this contains a `+08:00` which postgres can interpret properly.

[the remark this answers is not in the document]
Comment (namrataukirde) [reply app/services/subscriptions_reminder.rb]: :+1:

**Pull request merged**